sdk/ts/src/gateway/index.ts

```diff
@@ -0,0 +1,67 @@
+import { CID } from 'multiformats'
+import { CarReader } from '@ipld/car'
+import fetch from 'cross-fetch'
+import { verify_block } from '../utils'
+import { Block } from '@ipld/car/reader'
+
+export const DEFAULT_GATEWAY = 'https://gateway.ursa.earth'
+
+/// Gateway provides a simple interface to the Ursa Gateway API.
+export class Gateway {
+    constructor(url?: URL | string) {
+        if (url) {
+            const urlStr =
+                typeof url === 'string' ? url : url.toString().slice(0, -1)
+            if (urlStr.endsWith('/')) {
+                this.url = urlStr.slice(0, -1)
+            } else {
+                this.url = urlStr
+            }
+        }
+    }
+
+    url = 'https://gateway.ursa.earth'
+
+    // Fetch a car file with a given root cid from the gateway
+    async get(
+        cid: CID,
+        verify = true,
+        _origins?: [string]
+    ): Promise<Block[] | undefined> {
+        const res = await fetch(`${this.url}/${cid.toString()}`, {
+            method: 'GET',
+            headers: {
+                Accept: 'application/vnd.ipfs.car'
+            }
+        })
+
+        const carFile = await res.arrayBuffer()
+
+        // https://github.com/ipld/js-car/#async-carreaderfrombytesbytes
+        const reader = await CarReader.fromBytes(new Uint8Array(carFile))
+        const roots = (await reader.getRoots()).map((cid) => {
+            return cid.toString()
+        })
+
+        if (!roots.includes(cid.toString())) {
+            console.log(roots)
+            throw new Error('Root CID not found in car file')
+        }
+
+        const content = []
+        for await (const block of reader.blocks()) {
+            if (verify) {
+                try {
+                    await verify_block(block.cid, block.bytes)
```

Comment:
It's blocking per iteration, isn't it?

Reply:
Hmm, yeah. Do you think an improvement would be to promise.all a map and perform all at the same time async?

Reply:
maybe:
![image](https://user-images.githubusercontent.com/26348060/217366741-95ed5e6f-5247-4e94-a6f3-191e2668b8a3.png)

Reply:
perfect, merged :ok_hand: https://github.com/fleek-network/ursa/pull/370/commits/7ec1498443530217b006231903097bfbc9a90668